Add gzip to nginx conf

MonPote opened the pull request:
**Component**: ui

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 
Currently the browser need to download 3Mb of JS. We can drastically decrease this number by compression thanks to nginx.

**Summary**:

- Enable gzip

This will decrease the size of the file the client need to download
In the current version we have ~7x improvement.
From 3.17Mb to 0.470Mb send to the client.

**Acceptance criteria**: 

You can reach the UI and play with it. Everything should be the same.
When you load the UI for the first time, open the console before and on the network tab you should see for the JS files something like [this](https://puu.sh/EcU7g/4a22ffde50.png) instead of something like [this](https://puu.sh/EcU6m/9ff1243eed.png)

bert-e commented:
# Hello monpote,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

slaperche-scality commented on ui/conf/nginx.conf:
```diff
@@ -12,13 +28,14 @@ server {
         try_files $uri $uri/ /index.html;
     }
 
+    # Provide the ConfigMaps generated by MetalK8s
```
not related to gzip

slaperche-scality commented on ui/conf/nginx.conf:
```diff
@@ -12,13 +28,14 @@ server {
         alias /etc/metalk8s/ui/config.json;
     }
 
+    # Use the theme provided by the customer
```
ditto (should be in a dedicated commit "add comments" or something like that.

slaperche-scality commented on ui/conf/nginx.conf:
```diff
@@ -2,8 +2,24 @@ server {
 
-    #charset koi8-r;
-    #access_log  /var/log/nginx/host.access.log  main;
+    # https://www.digitalocean.com/community/tutorials/how-to-increase-pagespeed-score-by-changing-your-nginx-configuration-on-ubuntu-16-04
```
I would rather link the official nginx documentation here: https://nginx.org/en/docs/http/ngx_http_gzip_module.html instead of a "random" blog.

slaperche-scality commented on ui/conf/nginx.conf:
```diff
@@ -2,8 +2,24 @@ server {
+    gzip on;
+    gzip_comp_level    5;
+    gzip_min_length    256;
+    gzip_proxied       any;
```
Do we need this one?

slaperche-scality commented on ui/conf/nginx.conf:
```diff
@@ -2,8 +2,24 @@ server {
+    gzip_comp_level    5;
+    gzip_min_length    256;
+    gzip_proxied       any;
+    gzip_vary          on;
```
Ditto, do we need this.

Seems useful to handle a very rare case so not sure it's worth it.

MonPote committed:
Add explanation of nginx conf

MonPote committed:
Add compression-webpack-plugin to create gziped file durint the build

slaperche-scality commented on ui/config-overrides.js:
```diff
@@ -1,3 +1,13 @@
-const { useBabelRc, override, useEslintRc } = require('customize-cra');
+const {
```
Hum, wrong commit?

Looks like this change belongs to "Add compression-webpack-plugin to create gziped file durint the build" instead of "Add gzip to nginx conf".

MonPote committed:
Add gzip to nginx conf

This will decrease the size of the file the client need to download
In the current version we have ~7x improvement.
From 3.17Mb to 0.470Mb send to the client.

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/add-gzip-to-decrease-the-bundle-size-send-to-the-client` and the
destination branch `development/2.4`.

Please resolve the conflict on **the feature branch** (`improvement/add-gzip-to-decrease-the-bundle-size-send-to-the-client`).

```bash
 $ git fetch
 $ git checkout origin/improvement/add-gzip-to-decrease-the-bundle-size-send-to-the-client
 $ git merge origin/development/2.4
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/add-gzip-to-decrease-the-bundle-size-send-to-the-client
```

NicolasT commented on ui/conf/nginx.conf:
```diff
@@ -2,8 +2,23 @@ server {
+    gzip_min_length    256;
+
+    gzip_types
+    application/javascript
```
Consider indenting these a bit more, for clarity.

MonPote replied:
Yep, I just add an indent.

NicolasT commented on ui/config-overrides.js:
```diff
@@ -1,3 +1,13 @@
+module.exports = override(
+  useBabelRc(),
+  useEslintRc(),
+  addWebpackPlugin(new CompressionPlugin())
```
Add a `,` at the end so later commits won't change this line for no good reason.

MonPote replied:
I just add the rule on .prettierc that will also enforce that rule in the entire project :)

MonPote committed:
Improve readability of the code

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

NicolasT reviewed (approved):
OK. However:

- Could you file a ticket for code/bundle-splitting.
- Could you file a ticket for configuring Webpack's bundle-size-limits-becomes-a-build-error feature, so CI catches bundle growth accordingly?

Both of these could be handled in the same PR of course.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

MonPote commented:
@bert-e approve

NicolasT commented:
For reference, https://webpack.js.org/configuration/performance/#performancehints

MonPote commented:
> For reference, https://webpack.js.org/configuration/performance/#performancehints



> OK. However:
> 
> * Could you file a ticket for code/bundle-splitting.
> * Could you file a ticket for configuring Webpack's bundle-size-limits-becomes-a-build-error feature, so CI catches bundle growth accordingly?
> 
> Both of these could be handled in the same PR of course.

Ok let me do that.

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye monpote.
